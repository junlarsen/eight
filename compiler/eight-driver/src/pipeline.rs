use bumpalo::Bump;
use eight_hir::arena::HirArena;
use eight_hir::query::HirQueryDatabase;
use eight_hir::syntax_lowering_pass::ASTSyntaxLoweringPass;
use eight_hir::textual_pass::HirModuleDebugPass;
use eight_hir::type_check_pass::{HirModuleTypeCheckerPass, TypingContext};
use eight_syntax::arena::AstArena;

pub struct PipelineOptions {
    pub emit_ast: bool,
    pub emit_hir: bool,
}

pub struct Pipeline {
    options: PipelineOptions,
}

impl Pipeline {
    pub fn new(options: PipelineOptions) -> Self {
        Self { options }
    }

    pub fn run<T: AsRef<str>>(&self, source: T) -> miette::Result<()> {
        let ast_bump = Bump::new();
        let ast_arena = AstArena::new(&ast_bump);
        let mut lexer = eight_syntax::lexer::Lexer::new(source.as_ref());
        let mut parser = eight_syntax::parser::Parser::new(&mut lexer, &ast_arena);
        let tu = parser.parse()?;
        if self.options.emit_ast {
            let syntax = ron::ser::to_string_pretty(&tu, Default::default())
                .expect("failed to serialize ast to ron");
            println!("{}", syntax);
        }

        let hir_bump = Bump::new();
        let hir_arena = HirArena::new(&hir_bump);
        let lowering_pass = ASTSyntaxLoweringPass::new(&hir_arena);
        let mut module = lowering_pass.visit_translation_unit(&tu)?;
        let query_database = HirQueryDatabase::new(module.signature);
        let mut cx = TypingContext::new(&hir_arena, &query_database);
        HirModuleTypeCheckerPass::visit(&mut module, &mut cx)?;
        if self.options.emit_hir {
            let doc = HirModuleDebugPass::format_hir_module_to_string(&module);
            println!("{}", doc);
        }
        Ok(())
    }
}
