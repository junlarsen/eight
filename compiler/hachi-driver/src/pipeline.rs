use bumpalo::Bump;
use hachi_hir::passes::{
    ASTSyntaxLoweringPass, HirModuleDebugPass, HirModuleTypeCheckerPass, TypingContext,
};
use hachi_hir::ty::HirTyArena;

pub struct PipelineOptions {
    pub emit_ast: bool,
    pub emit_hir: bool,
}

pub struct Pipeline {
    options: PipelineOptions,
}

impl Pipeline {
    pub fn new(options: PipelineOptions) -> Self {
        Self { options }
    }

    pub fn run<T: AsRef<str>>(&self, source: T) -> miette::Result<()> {
        let mut lexer = hachi_syntax::Lexer::new(source.as_ref());
        let mut parser = hachi_syntax::Parser::new(&mut lexer);

        let bump = Bump::new();
        let arena = HirTyArena::new(&bump);

        let mut lowering_pass = ASTSyntaxLoweringPass::new(&arena);
        let tu = parser.parse()?;
        let mut module = lowering_pass.visit_translation_unit(&tu)?;
        let mut cx = TypingContext::new(&arena);
        HirModuleTypeCheckerPass::visit(&mut module, &mut cx)?;
        if self.options.emit_ast {
            let syntax = ron::ser::to_string_pretty(&tu, Default::default())
                .expect("failed to serialize ast to ron");
            println!("{}", syntax);
        }
        if self.options.emit_hir {
            let doc = HirModuleDebugPass::format_hir_module_to_string(&module);
            println!("{}", doc);
        }
        Ok(())
    }
}
