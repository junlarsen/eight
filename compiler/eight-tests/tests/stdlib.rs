use eight_driver::pipeline::{Pipeline, PipelineOptions};
use insta_cmd::get_cargo_bin;
use std::path::Path;
use std::process::Command;

#[test]
fn test_evaluate_stdlib() {
    let path = Path::join(env!("CARGO_MANIFEST_DIR").as_ref(), "../../stdlib");
    let files = std::fs::read_dir(path)
        .expect("failed to read stdlib directory")
        .filter_map(|f| f.ok())
        .filter(|f| {
            std::fs::metadata(f.path())
                .expect("failed to stat file")
                .is_file()
        });
    let mut buf = String::new();
    for file in files {
        let path = file.path();
        let content = std::fs::read_to_string(path).expect("failed to read file");
        buf.push_str(&content);
        buf.push('\n');
    }

    let mut cmd = Command::new(get_cargo_bin("eightc"));
    insta_cmd::assert_cmd_snapshot!(cmd.arg("-").arg("--emit-hir").pass_stdin(buf));
}
