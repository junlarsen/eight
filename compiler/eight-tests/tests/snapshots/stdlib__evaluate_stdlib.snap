---
source: compiler/eight-tests/tests/stdlib.rs
info:
  program: eightc
  args:
    - "-"
    - "--emit-hir"
  stdin: "intrinsic_fn malloc<T>(size: i32) -> *T;\nintrinsic_fn free<T>(ptr: *T) -> unit;\nintrinsic_fn exit() -> unit;\n\nintrinsic_type bool;\n\nintrinsic_fn __builtin_bool_and(a: bool, b: bool) -> bool;\nintrinsic_fn __builtin_bool_or(a: bool, b: bool) -> bool;\n\ninstance And<bool, bool> {\n  fn and(a: bool, b: bool) -> bool {\n    return __builtin_bool_and(a, b);\n  }\n}\ninstance Or<bool, bool> {\n  fn or(a: bool, b: bool) -> bool {\n    return __builtin_bool_or(a, b);\n  }\n}\n\nintrinsic_fn __builtin_bool_eq(a: bool, b: bool) -> bool;\nintrinsic_fn __builtin_bool_neq(a: bool, b: bool) -> bool;\n\ninstance Eq<bool, bool> {\n  fn eq(a: bool, b: bool) -> bool {\n    return __builtin_bool_eq(a, b);\n  }\n  fn neq(a: bool, b: bool) -> bool {\n    return __builtin_bool_neq(a, b);\n  }\n}\n\nintrinsic_fn __builtin_bool_not(a: bool) -> bool;\n\ninstance Not<bool, bool> {\n  fn not(a: bool) -> bool {\n    return __builtin_bool_not(a);\n  }\n}\n\nintrinsic_type i32;\n\nintrinsic_fn __builtin_i32_add(a: i32, b: i32) -> i32;\nintrinsic_fn __builtin_i32_sub(a: i32, b: i32) -> i32;\nintrinsic_fn __builtin_i32_mul(a: i32, b: i32) -> i32;\nintrinsic_fn __builtin_i32_div(a: i32, b: i32) -> i32;\nintrinsic_fn __builtin_i32_rem(a: i32, b: i32) -> i32;\n\ninstance Add<i32, i32, i32> {\n  fn add(a: i32, b: i32) -> i32 {\n    return __builtin_i32_add(a, b);\n  }\n}\n\ninstance Sub<i32, i32, i32> {\n  fn sub(a: i32, b: i32) -> i32 {\n    return __builtin_i32_sub(a, b);\n  }\n}\n\ninstance Mul<i32, i32, i32> {\n  fn mul(a: i32, b: i32) -> i32 {\n    return __builtin_i32_mul(a, b);\n  }\n}\n\ninstance Div<i32, i32, i32> {\n  fn div(a: i32, b: i32) -> i32 {\n    return __builtin_i32_div(a, b);\n  }\n}\n\ninstance Rem<i32, i32, i32> {\n  fn rem(a: i32, b: i32) -> i32 {\n    return __builtin_i32_rem(a, b);\n  }\n}\n\nintrinsic_fn __builtin_i32_eq(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_neq(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_lt(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_gt(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_le(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_ge(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_lte(a: i32, b: i32) -> bool;\nintrinsic_fn __builtin_i32_gte(a: i32, b: i32) -> bool;\n\ninstance Eq<i32, i32> {\n  fn eq(a: i32, b: i32) -> bool {\n    return __builtin_i32_eq(a, b);\n  }\n  fn neq(a: i32, b: i32) -> bool {\n    return __builtin_i32_neq(a, b);\n  }\n}\n\ninstance Ord<i32, i32> {\n  fn lt(a: i32, b: i32) -> bool {\n    return __builtin_i32_lt(a, b);\n  }\n  fn gt(a: i32, b: i32) -> bool {\n    return __builtin_i32_gt(a, b);\n  }\n  fn le(a: i32, b: i32) -> bool {\n    return __builtin_i32_le(a, b);\n  }\n  fn ge(a: i32, b: i32) -> bool {\n    return __builtin_i32_ge(a, b);\n  }\n}\n\nintrinsic_fn __builtin_i32_neg(a: i32) -> i32;\n\ninstance Neg<i32, i32> {\n  fn neg(a: i32) -> i32 {\n    return __builtin_i32_neg(a);\n  }\n}\n\nintrinsic_type unit;\n\ntrait Add<A, B, R> {\n  fn add(a: A, b: B) -> R;\n}\n\ntrait Sub<A, B, R> {\n  fn sub(a: A, b: B) -> R;\n}\n\ntrait Mul<A, B, R> {\n  fn mul(a: A, b: B) -> R;\n}\n\ntrait Div<A, B, R> {\n  fn div(a: A, b: B) -> R;\n}\n\ntrait Rem<A, B, R> {\n  fn rem(a: A, b: B) -> R;\n}\n\ntrait Eq<A, B> {\n  fn eq(a: A, b: B) -> bool;\n  fn neq(a: A, b: B) -> bool;\n}\n\ntrait Ord<A, B> {\n  fn lt(a: A, b: B) -> i32;\n  fn gt(a: A, b: B) -> i32;\n  fn le(a: A, b: B) -> i32;\n  fn ge(a: A, b: B) -> i32;\n}\n\ntrait Neg<A, R> {\n  fn neg(a: A) -> R;\n}\n\ntrait Not<A, R> {\n  fn not(a: A) -> R;\n}\n\ntrait And<A, B> {\n  fn and(a: A, b: B) -> bool;\n}\n\ntrait Or<A, B> {\n  fn or(a: A, b: B) -> bool;\n}\n\n"
snapshot_kind: text
---
success: true
exit_code: 0
----- stdout -----
hir_module {
  declares {
    // module types
    type bool;
    type i32;
    type unit;
    
    // module struct types
    
    
    // module functions
    fn __builtin_bool_and(a: bool, b: bool) -> bool;
    fn __builtin_bool_eq(a: bool, b: bool) -> bool;
    fn __builtin_bool_neq(a: bool, b: bool) -> bool;
    fn __builtin_bool_not(a: bool) -> bool;
    fn __builtin_bool_or(a: bool, b: bool) -> bool;
    fn __builtin_i32_add(a: i32, b: i32) -> i32;
    fn __builtin_i32_div(a: i32, b: i32) -> i32;
    fn __builtin_i32_eq(a: i32, b: i32) -> bool;
    fn __builtin_i32_ge(a: i32, b: i32) -> bool;
    fn __builtin_i32_gt(a: i32, b: i32) -> bool;
    fn __builtin_i32_gte(a: i32, b: i32) -> bool;
    fn __builtin_i32_le(a: i32, b: i32) -> bool;
    fn __builtin_i32_lt(a: i32, b: i32) -> bool;
    fn __builtin_i32_lte(a: i32, b: i32) -> bool;
    fn __builtin_i32_mul(a: i32, b: i32) -> i32;
    fn __builtin_i32_neg(a: i32) -> i32;
    fn __builtin_i32_neq(a: i32, b: i32) -> bool;
    fn __builtin_i32_rem(a: i32, b: i32) -> i32;
    fn __builtin_i32_sub(a: i32, b: i32) -> i32;
    fn exit() -> unit;
    fn free<T>(ptr: *T) -> unit;
    fn malloc<T>(size: i32) -> *T;
    
    // module traits
    trait Add<A, B, R> {
      fn add(a: A, b: B) -> R;
    }
    trait And<A, B> {
      fn and(a: A, b: B) -> bool;
    }
    trait Div<A, B, R> {
      fn div(a: A, b: B) -> R;
    }
    trait Eq<A, B> {
      fn eq(a: A, b: B) -> bool;
      fn neq(a: A, b: B) -> bool;
    }
    trait Mul<A, B, R> {
      fn mul(a: A, b: B) -> R;
    }
    trait Neg<A, R> {
      fn neg(a: A) -> R;
    }
    trait Not<A, R> {
      fn not(a: A) -> R;
    }
    trait Or<A, B> {
      fn or(a: A, b: B) -> bool;
    }
    trait Ord<A, B> {
      fn ge(a: A, b: B) -> i32;
      fn gt(a: A, b: B) -> i32;
      fn le(a: A, b: B) -> i32;
      fn lt(a: A, b: B) -> i32;
    }
    trait Rem<A, B, R> {
      fn rem(a: A, b: B) -> R;
    }
    trait Sub<A, B, R> {
      fn sub(a: A, b: B) -> R;
    }
    
    // module instances
    instance And<bool, bool> {
      fn and(a: bool, b: bool) -> bool;
    }
    instance Or<bool, bool> {
      fn or(a: bool, b: bool) -> bool;
    }
    instance Eq<bool, bool> {
      fn eq(a: bool, b: bool) -> bool;
      fn neq(a: bool, b: bool) -> bool;
    }
    instance Not<bool, bool> {
      fn not(a: bool) -> bool;
    }
    instance Add<i32, i32, i32> {
      fn add(a: i32, b: i32) -> i32;
    }
    instance Sub<i32, i32, i32> {
      fn sub(a: i32, b: i32) -> i32;
    }
    instance Mul<i32, i32, i32> {
      fn mul(a: i32, b: i32) -> i32;
    }
    instance Div<i32, i32, i32> {
      fn div(a: i32, b: i32) -> i32;
    }
    instance Rem<i32, i32, i32> {
      fn rem(a: i32, b: i32) -> i32;
    }
    instance Eq<i32, i32> {
      fn eq(a: i32, b: i32) -> bool;
      fn neq(a: i32, b: i32) -> bool;
    }
    instance Ord<i32, i32> {
      fn ge(a: i32, b: i32) -> bool;
      fn gt(a: i32, b: i32) -> bool;
      fn le(a: i32, b: i32) -> bool;
      fn lt(a: i32, b: i32) -> bool;
    }
    instance Neg<i32, i32> {
      fn neg(a: i32) -> i32;
    }
  }
  
  defining {
    // module functions
    
    // module intrinsic functions
    instance And<bool, bool> {
      fn and<>(a: bool, b: bool) -> bool {
        return ((__builtin_bool_and as fn((bool, bool))->bool)::<>((a as bool),(b as bool)) as bool);
      }
    }
    instance Or<bool, bool> {
      fn or<>(a: bool, b: bool) -> bool {
        return ((__builtin_bool_or as fn((bool, bool))->bool)::<>((a as bool),(b as bool)) as bool);
      }
    }
    instance Eq<bool, bool> {
      fn eq<>(a: bool, b: bool) -> bool {
        return ((__builtin_bool_eq as fn((bool, bool))->bool)::<>((a as bool),(b as bool)) as bool);
      }
      fn neq<>(a: bool, b: bool) -> bool {
        return ((__builtin_bool_neq as fn((bool, bool))->bool)::<>((a as bool),(b as bool)) as bool);
      }
    }
    instance Not<bool, bool> {
      fn not<>(a: bool) -> bool {
        return ((__builtin_bool_not as fn((bool))->bool)::<>((a as bool)) as bool);
      }
    }
    instance Add<i32, i32, i32> {
      fn add<>(a: i32, b: i32) -> i32 {
        return ((__builtin_i32_add as fn((i32, i32))->i32)::<>((a as i32),(b as i32)) as i32);
      }
    }
    instance Sub<i32, i32, i32> {
      fn sub<>(a: i32, b: i32) -> i32 {
        return ((__builtin_i32_sub as fn((i32, i32))->i32)::<>((a as i32),(b as i32)) as i32);
      }
    }
    instance Mul<i32, i32, i32> {
      fn mul<>(a: i32, b: i32) -> i32 {
        return ((__builtin_i32_mul as fn((i32, i32))->i32)::<>((a as i32),(b as i32)) as i32);
      }
    }
    instance Div<i32, i32, i32> {
      fn div<>(a: i32, b: i32) -> i32 {
        return ((__builtin_i32_div as fn((i32, i32))->i32)::<>((a as i32),(b as i32)) as i32);
      }
    }
    instance Rem<i32, i32, i32> {
      fn rem<>(a: i32, b: i32) -> i32 {
        return ((__builtin_i32_rem as fn((i32, i32))->i32)::<>((a as i32),(b as i32)) as i32);
      }
    }
    instance Eq<i32, i32> {
      fn eq<>(a: i32, b: i32) -> bool {
        return ((__builtin_i32_eq as fn((i32, i32))->bool)::<>((a as i32),(b as i32)) as bool);
      }
      fn neq<>(a: i32, b: i32) -> bool {
        return ((__builtin_i32_neq as fn((i32, i32))->bool)::<>((a as i32),(b as i32)) as bool);
      }
    }
    instance Ord<i32, i32> {
      fn lt<>(a: i32, b: i32) -> bool {
        return ((__builtin_i32_lt as fn((i32, i32))->bool)::<>((a as i32),(b as i32)) as bool);
      }
      fn gt<>(a: i32, b: i32) -> bool {
        return ((__builtin_i32_gt as fn((i32, i32))->bool)::<>((a as i32),(b as i32)) as bool);
      }
      fn le<>(a: i32, b: i32) -> bool {
        return ((__builtin_i32_le as fn((i32, i32))->bool)::<>((a as i32),(b as i32)) as bool);
      }
      fn ge<>(a: i32, b: i32) -> bool {
        return ((__builtin_i32_ge as fn((i32, i32))->bool)::<>((a as i32),(b as i32)) as bool);
      }
    }
    instance Neg<i32, i32> {
      fn neg<>(a: i32) -> i32 {
        return ((__builtin_i32_neg as fn((i32))->i32)::<>((a as i32)) as i32);
      }
    }
  
  }
}

----- stderr -----
